### ISSUE TITLE
Arangosh faulty behavior with CTRL-C

### ISSUE THREAD

**Issue by lewisdiamond** [arangodb/arangodb#2176]
## my environment running ArangoDB
I'm using the latest ArangoDB of the respective release series:
- [ ] 2.8
- [x] 3.0
- [ ] self-compiled devel branch

On this operating system:
- [ ] DCOS on
  - [ ] AWS
  - [ ] Azure
  - [ ] own infrastructure
- [x] Linux 
  - [ ] Debian .deb
  - [ ] Ubuntu .deb
  - [ ] SUSE   .rpm
  - [ ] RedHat .rpm
  - [ ] Fedora .rpm
  - [ ] Gentoo
  - [x] docker - official docker library
  - [ ] other:
- [ ] Windows, version: 
- [ ] MacOS, version:

### this is an AQL-related issue:
[ ] I'm using graph features

I'm issuing AQL via:
- [ ] web interface with this browser:     running on this OS:
- [x] arangosh
- [ ] this Driver:

# CTRL+C
Expected: In arangosh, start typing on the input line, CTRL+C gives you a new line without executing the input entered, press CTRL+C again, prompt "Press CTRL+C again to exit", CTRL+C again exits (Same behavior as the nodejs shell). 

Actual: In arangosh, start typing on the input line, hit CTRL+C and arangosh exits completely. This behavior would be expected of CTRL+D not CTRL+C (the exit message even displays "CTRL+D")

**Comment by jsteemann** [comment]
Suggested fix for 3.0:
```diff
diff --git a/3rdParty/linenoise-ng/include/linenoise.h b/3rdParty/linenoise-ng/include/linenoise.h
index aac6282..3ca338d 100644
--- a/3rdParty/linenoise-ng/include/linenoise.h
+++ b/3rdParty/linenoise-ng/include/linenoise.h
@@ -63,6 +63,7 @@ void linenoiseSetMultiLine(int ml);
 void linenoisePrintKeyCodes(void);
 /* the following is extension to the original linenoise API */
 int linenoiseInstallWindowChangeHandler(void);
+int linenoiseGotKey(void);
 
 #ifdef __cplusplus
 }
diff --git a/3rdParty/linenoise-ng/src/linenoise.cpp b/3rdParty/linenoise-ng/src/linenoise.cpp
index 3c68701..06689f6 100644
--- a/3rdParty/linenoise-ng/src/linenoise.cpp
+++ b/3rdParty/linenoise-ng/src/linenoise.cpp
@@ -2483,8 +2483,11 @@ static bool isCharacterAlphanumeric(char32_t testChar) {
 #ifndef _WIN32
 static bool gotResize = false;
 #endif
+static int gotKey = 0;
 
 int InputBuffer::getInputLine(PromptBase& pi) {
+  gotKey = 0;
+
   // The latest history entry is always our current buffer
   if (len > 0) {
     size_t bufferSize = sizeof(char32_t) * len + 1;
@@ -2527,6 +2530,11 @@ int InputBuffer::getInputLine(PromptBase& pi) {
     if (terminatingKeystroke == -1) {
       c = linenoiseReadChar();  // get a new keystroke
 
+      if (c != 0 && (c != ctrlChar('C'))) {
+        // set flag that we got some input
+        gotKey = 1;
+      }
+
 #ifndef _WIN32
       if (c == 0 && gotResize) {
         // caught a window resize event
@@ -3413,3 +3421,7 @@ int linenoiseInstallWindowChangeHandler(void) {
 #endif
   return 0;
 }
+
+int linenoiseGotKey(void) {
+  return gotKey;
+}
diff --git a/arangosh/Shell/V8ShellFeature.cpp b/arangosh/Shell/V8ShellFeature.cpp
index 6a00d56..d2840c7 100644
--- a/arangosh/Shell/V8ShellFeature.cpp
+++ b/arangosh/Shell/V8ShellFeature.cpp
@@ -311,13 +311,14 @@ int V8ShellFeature::runShell(std::vector<std::string> const& positionals) {
 
   uint64_t nrCommands = 0;
 
-  ClientFeature* client =
-      dynamic_cast<ClientFeature*>(server()->feature("Client"));
+  ClientFeature* client = server()->getFeature<ClientFeature>("Client");
 
   if (!client->isEnabled()) {
     client = nullptr;
   }
 
+  bool lastEmpty = false;
+
   while (true) {
     _console->setPromptError(promptError);
     auto prompt = _console->buildPrompt(client);
@@ -326,14 +327,16 @@ int V8ShellFeature::runShell(std::vector<std::string> const& positionals) {
     std::string input =
         v8LineEditor.prompt(prompt._colored, prompt._plain, eof);
 
-    if (eof) {
+    if (eof && lastEmpty) {
       break;
     }
 
     if (input.empty()) {
       promptError = false;
+      lastEmpty = true;
       continue;
     }
+    lastEmpty = false;
 
     _console->log(prompt._plain + input + "\n");
 
@@ -406,7 +409,7 @@ int V8ShellFeature::runShell(std::vector<std::string> const& positionals) {
       TRI_RunGarbageCollectionV8(_isolate, 500.0);
     }
   }
-
+      
   if (!_console->quiet()) {
     _console->printLine("");
     _console->printByeBye();
diff --git a/lib/Utilities/LinenoiseShell.cpp b/lib/Utilities/LinenoiseShell.cpp
index d401b8a..1595ae6 100644
--- a/lib/Utilities/LinenoiseShell.cpp
+++ b/lib/Utilities/LinenoiseShell.cpp
@@ -99,11 +99,15 @@ std::string LinenoiseShell::getLine(std::string const& prompt, bool& eof) {
 
   if (line != nullptr) {
     eof = false;
-    std::string const stringValue(line);
+    std::string stringValue(line);
     ::free(line);
     return stringValue;
   }
 
-  eof = true;
+  if (!linenoiseGotKey() || !isatty(STDIN_FILENO)) {
+    // only set eof if we did not get any key presses from linenoise
+    eof = true;
+  }
+
   return "";
 }
```

**Comment by lewisdiamond** [comment]
Well that was fast.

**Comment by jsteemann** [comment]
Fixed in 3.0 (3.0.12), 3.1 (3.1.2) and devel.

**Comment by dothebart** [comment]
3.0.12 is available for download via the regular upgrade channels - closing.
